feat(authz/config): Add a pipeline roles section to pipeline config.

dibyom opened the pull request:
Deck part of https://github.com/spinnaker/spinnaker/issues/3037

Here is what this looks like:
![screenshot from 2018-07-20 15-58-46](https://user-images.githubusercontent.com/3316877/43022623-daf897de-8c35-11e8-80da-b1d5f5e597cc.png)

danielpeach commented on app/scripts/modules/core/src/help/help.contents.ts:
```diff
@@ -402,6 +402,9 @@ const helpContents: { [key: string]: string } = {
       when triggering the pipeline manually. You can include HTML in this field.`,
   'pipeline.config.failOnFailedExpressions': `When this option is enabled, the stage will be marked as failed if it contains any failed expressions`,
+  'pipeline.config.roles.help': `
+  <p>The roles specified will be used for making authorization decisions when the pipeline is triggerd using an automated trigger.
```
Please indent here.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/pipelineRoles.component.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  public toggleCheck(role: string) {
+    if (this.pipeline.roles.indexOf(role) === -1) {
```
It's a little more idiomatic to use `if (this.pipeline.roles.includes(role))`, though the boolean is flipped in that case.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/pipelineRoles.component.ts:
```diff
@@ -0,0 +1,46 @@
+      <div class="row checkbox" ng-repeat="role in permissionCtrl.allowedRoles">
+        <label>
+          <input type="checkbox"
+          ng-checked="permissionCtrl.pipeline.roles.indexOf(role) > -1"
```
Please indent this line and the line below.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/runAsUserSelector.component.ts:
```diff
@@ -1,25 +1,38 @@
+  public enabled = true;
+
+  public $onInit(): void {
+    this.enabled = SETTINGS.feature['fiatEnabled'] && !SETTINGS.feature['managedServiceAccounts'];
```
You don't need the bracket notation - can just get the key `SETTINGS.feature.managedServiceAccounts`. It's possible that the TS compiler will complain, but in that case you should add `managedServiceaccounts` to `settings.ts`, which types `settings.js`.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/pipelineRoles.component.ts:
```diff
@@ -0,0 +1,46 @@
+          <input type="checkbox"
+          ng-checked="permissionCtrl.pipeline.roles.indexOf(role) > -1"
+          ng-click="permissionCtrl.toggleCheck(role)">
+          <strong>{{role}}</strong>
```
I think this might be better as a multiselect rather than a checkbox list. If a user has many roles (maybe > 10), this list will become unwieldy. Do a search for `<ui-select>` in this repo - any with the attribute `multiple` should be a good example. The multiselect will also allow users to search for their roles.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/triggers.html:
```diff
@@ -18,6 +18,17 @@
     </div>
   </page-section>
+  <render-if-feature feature="managedServiceAccounts">
+  <page-section key="roles" label="Pipeline Roles" visible="!pipeline.strategy" no-wrapper="true">
```
Please indent this section between the `render-if-feature` tags.

danielpeach replied:
I think this would be better positioned under `Automated Triggers` below, rather than as its own page section - what do you think? I'd put it at the top of the trigger configs.

danielpeach commented on app/scripts/modules/core/src/config/settings.ts:
```diff
@@ -47,6 +47,7 @@ export interface IFeatures {
   pipelineTemplates?: boolean;
   versionedProviders?: boolean;
   travis?: boolean;
+  managedServiceAccountsEnabled?: boolean;
```
This doesn't match the `settings.js` value you've configured - I think it should be `managedServiceAccounts`.

dibyom commented:
Changed the roles selector to a multi-select and moved it under `Automated Triggers`. Only shows up if a trigger has been added:
![screenshot from 2018-07-23 13-20-12](https://user-images.githubusercontent.com/3316877/43092467-4e7eb70a-8e7b-11e8-8184-f12335521c2d.png)

Also reworded the help text to make it clearer.
PTAL.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/pipelineRoles.component.ts:
```diff
@@ -0,0 +1,48 @@
+  }
+}
+
+class PipelineRolesComponent implements ng.IComponentOptions {
```
Ideally everything would be pipeline "permissions" or pipeline "roles" - but I am ok with that happening in a different PR.
